sklearn/preprocessing/_label.py

```diff
@@ -111,7 +111,8 @@ def _encode(values, uniques=None, encode=False, check_unknown=True):
         try:
             res = _encode_python(values, uniques, encode)
         except TypeError:
-            raise TypeError("argument must be a string or number")
+            raise TypeError("Encoders require their input to be strings or "
+                            "numbers.")
```

Comment:
I would like to make the error message as informative as possible and report the observed types:

```suggestion
            types = set(type(v) for v in values)
            raise TypeError("Encoders require their input to be uniformly "
                            f"strings or numbers. Got {types}")
```

Note that the order of sets is not deterministic so it would not be possible to change the test to also check the observed types in 

One could make the order deterministic with something like:

```python
type_names = sorted(t.__qualname__ for t in set(type(v) for v in values))
```
but I am not sure that `__qualname__` is always defined.

Using `str` or `repr`as follows:

```python
sorted(str(t) for t in set(type(v) for v in values))
```

both lead to ugly outputs such as:

```
["<class 'NoneType'>", "<class 'float'>", "<class 'str'>"]
```

instead of the short names from `__qualname__`:

```
['NoneType', 'float', 'str']
```

Reply:
We can try to use `__qualname__` and if an `AttributeError` is raised we can always go back to the type, but that would produce different results depending on the input.

However, I think __qualname__ was introduced in [python 3.3,](https://docs.python.org/3/glossary.html#term-qualified-name) so it should be supported

Reply:
Yes `__qualname__` should always be there but maybe we overlook something. But let's do `__qualname__`. This code branch is only executed in case of a `TypeError` so this change will not break code that already work without error.